### ISSUE TITLE
[Discussion] Adding virt-who plugin cases to satellite jenkins job

### ISSUE THREAD

**Issue by eko999** [SatelliteQE/robottelo#7746]
Hey all, I'd like to start a discussion about adding virt-who plugin cases to satellite jenkins job, in the early days, we started to design the test cases for virt-who plugin as the issue https://github.com/SatelliteQE/robottelo/issues/7006:
1). Virt-who plugin test cases will be managed in robottelo/tests/foreman/virtwho
2). The test cases will be run by virt-who jenkins because we need to provision the complex hypervisors by virt-who jenkins.

And now virt-who plugin cases can be run normally by https://virtwho-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/Virtwho-Configure/job/Virtwho-Configure-Pipeline/, but we realized virt-who plugin is one component of satellite product, there is no reason we use different jobs to run virt-who plugin testing, virt-who plugin's quality should be measured with the whole satellite product, so we hope virt-who plugin cases can be run and triggered by satellite jenkins jobs.

In order to make this possible, we should finish the following steps firstly:
1). Provide a configure file to include all the supported hypervisors configuration and then the job can read the configuration for each hypervisor.
2). Reduce the time of virt-who plugin testing, the currently 12 hours is not acceptable for satellite trigger job, we should optimize virt-who plugin test cases.

Do you guys have any concern or suggestion about this action, please feel free to let me know what's your opinion. Thanks.

**Comment by eko999** [comment]
@mirekdlugosz @vijay8451 @jyejare @lpramuk :^^

**Comment by mirekdlugosz** [comment]
This is complex topic, and we probably don't want to discuss some of details publicly in GitHub.

Overall, we lack coherent strategy towards plugins testing. Model that I see as naturally emerging is: "let Satellite QE test Satellite integration bit, let other team test rest of it". But then, I don't know what exactly that means for virt-who.

1. Is virt-who even a standalone offering?
2. For each interesting test scenario, can you reasonably argue whether it should belong to Satellite or virth-who?

If answer to first question is "no" (virt-who really only makes sense through Satellite), then yes, we should totally move execution to SatQE jenkins. If answer is "yes" (customer can use virt-who without Satellite), then we have to tackle second question. And I guess answer will be messy, as some scenarios can't be clearly attributes to single team.

> 1). Provide a configure file to include all the supported hypervisors configuration and then the job can read the configuration for each hypervisor.

Does jenkins job do anything special with these configs? Or are they simple robottelo.properties config files that we should incorporate into existing infra?

> 2). Reduce the time of virt-who plugin testing, the currently 12 hours is not acceptable for satellite trigger job, we should optimize virt-who plugin test cases.

I don't follow that closely, but I believe our automation takes longer than that. As long as we can run virt-who in parallel to other jobs, and preferably on separate Satellite instance (which should be easy-ish with VMbroker and Ansible Tower), I'd say 12 hours is acceptable.
I mean, of course any and all performance improvements are welcome, but that sounds like easiest problem here. Personally I wouldn't block entire process on that single item.

**Comment by mshriver** [comment]
Well said @mirekdlugosz, and thanks for starting the discussion @eko999.

I think that the resolution to #7006 still stands, and regardless of the changes to CI structure, we'll continue to maintain virt-who integration tests in the robottelo repo.

@mirekdlugosz that should answer your questions about it being a separate offering - while there may be virt-who specific tests that live outside of robottelo, the tests defined in robottelo are for the integration with foreman/Satellite. @eko999 and the virt-who testers are part of the SatQE team, reporting results together.

The discussion of format for configuration files concerning virt-who hypervisor targets is relevant to robottelo, but is in some way dictated by the planned work to move robottelo to dynaconf for settings reading. 

Please use #6756 (recently re-named and re-assigned) to discuss the config file requirements for defining virt-who hypervisors for the relevant test cases.

I would like to continue discussion of the structure of the CI jobs, and the scope of their test collection and execution environments, on internal communication channels.

**Comment by eko999** [comment]
@mirekdlugosz 

> 1. Is virt-who even a standalone offering?
> 2. For each interesting test scenario, can you reasonably argue whether it should belong to Satellite or virth-who?

The answer of first question is Yes, virt-who package is shipped with RHEL7/8 and  @Junefen is responsible for the package's function testing.
But about the virt-who plugin of satellite, the reason I think it should be one part of satellite is:
1). The basic packages of the virt-who plugin are delivered with satellite product always.
```
# rpm -qa|grep virt_who
tfm-rubygem-hammer_cli_foreman_virt_who_configure-0.0.5.1-1.el7sat.noarch
tfm-rubygem-foreman_virt_who_configure-0.5.2-1.el7sat.noarch
```
2). The developer of the plugin is Marek Hulán, he is one member of satellite dev team.
3). Many customers use the plugin to create virt-who configure file instead of manual creating.

> Does jenkins job do anything special with these configs? Or are they simple robottelo.properties config files that we should incorporate into existing infra?
The jenkins job should loop each hypervisor type and then read the hypervisor configuration to update robottelo.properties, just like:
```
for hypervisor_type in ['esx', 'hyperv','xen','rhevm','libvirt','kubevirt]:
    read_hypervisor_configure_file(hypervisor_type)
    update_robottelo_properties(hypervisor_type)
    run_all_testcases()
```
All the cases will be run for each hypervisor, this is why we need almost 12h, but actually some of the cases are not necessary to be run for each hypervisor, so I'm thinking to optimize the cases as well.

**Comment by eko999** [comment]
@mshriver Thank you for raising up the issue in the function groups.
> The discussion of format for configuration files concerning virt-who hypervisor targets is relevant to robottelo, but is in some way dictated by the planned work to move robottelo to dynaconf for settings reading.

The configuration of virt-who was supported in robottelo.properties already:
```
# Section for virtwho configure function
# [virtwho]
# The hypervisor type, one of the values: esx, xen, hyperv, rhevm, libvirt, kubevirt
# hypervisor_type=
# The hostname or ip or other values for hypervisor server, depends on the type
# hypervisor_server=
# Admin username to access hypervisor server
# hypervisor_username=
# Admin password to access hypervisor server
# hypervisor_password=
# The config file to connect hypervisor, only for kubevirt
# hypervisor_config_file=
# The guest ip or hostname in the hypervisor
# guest=
# The guest port for ssh connection, by default guest_port=22
# guest_port=22
# ssh username to access guest
# guest_username=
# ssh password to access guest
# guest_password=
# The SKU ID for physical vdc
# sku_vdc_physical=
# The SKU ID for virtual vdc
# sku_vdc_virtual=
```
Of course, if we need to use dynnconf to instead of INI, I think we are in the same page as other components.

**Comment by hkx303** [comment]
This topic is about virt-who plugin cases, I will working on this as I am the component owner, thanks @eko999 for raising this and the detailed explaination,  thanks for the thoughts @mirekdlugosz @mshriver . Around the Needed changes, I have added some comments in Confluence page.

**Comment by mirekdlugosz** [comment]
> The answer of first question is Yes, virt-who package is shipped with RHEL7/8 and @Junefen is responsible for the package's function testing.
> But about the virt-who plugin of satellite, the reason I think it should be one part of satellite is:

These are all good arguments to make virth-who part of Satellite execution environment.

My intention was not to say we shouldn't do it. More to think about general design and tradeoffs we have to make. Example: hypervisor configuration files might need to be duplicated, one copy in Satellite jenkins-configs, one copy in repo used for RPM package testing.

 > The jenkins job should loop each hypervisor type and then read the hypervisor configuration to update robottelo.properties, just like:
> 
> ```
> for hypervisor_type in ['esx', 'hyperv','xen','rhevm','libvirt','kubevirt]:
>     read_hypervisor_configure_file(hypervisor_type)
>     update_robottelo_properties(hypervisor_type)
>     run_all_testcases()
> ```

I'm not 100% sure on that, but I believe dynaconf will include/allow switch to YAML config files, and then you would be able to use lists of config sections. Switch to dynaconf should at least make it easier to store all hypervisor configs in single file. And then your tests could be parametrized, so you could have single test code running for each hypervisor. All without any extra steps on Jenkins side.

Overall, this is direction I would be heading here. A lot of tests would benefit from capability like that.

**Comment by hkx303** [comment]
> I'm not 100% sure on that, but I believe dynaconf will include/allow switch to YAML config files, and then you would be able to use lists of config sections. Switch to dynaconf should at least make it easier to store all hypervisor configs in single file. And then your tests could be parametrized, so you could have single test code running for each hypervisor. All without any extra steps on Jenkins side.

That's great, any plan for switching to Dynaconf? Do we have some tickets to follow this :)

**Comment by mirekdlugosz** [comment]
@hkx303 you can take a look at jira card SATQE-9153, and "Automation Refactoring Goals" article on docs.engineering.
@rochacbruno leads this effort, he will be able to answer any more detailed questions.

There will also be tech talk on Thursday on Dynaconf. I think it will be more generic, I don't expect "porting robottelo to dynaconf" to be on list of discussed topics.

**Comment by mshriver** [comment]
@hkx303 I linked the robottelo issue above for the move to dynaconf, #6756.

@mirekdlugosz completely agree on the approach to storing necessary hypervisor configurations in yaml, and parametrizing the tests on the hypervisor type.  That's how I envisioned it, and I think that's a better approach than trying to re-write the settings content at runtime based on execution environment.

I think virt-who will have its own testing concerns, and custom configuration, for any package testing that isn't related to satellite integration. I don't know that we'll be able to eliminate duplication of the configs between the projects upstream, because they'll inherently diverge from our downstream integration testing.
